DataDog/integrations-core
# [Kubelet] Support Ephemeral Percistent volume claim

### What does this PR do?

This PR adds tags to ephemeral volume metrics. It leverages the fact that a standalone `pvc` is created and its name is formatted as `<pod_name>-<volume_name>`

### Motivation

In kubernetes 1.25 it is now possible to define a VolumeClaim template for ephemeral storage. This feature is call [generic ephemeral volume](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes).
The current code does not make the mapping correctly to find the associated pvc name.

### Additional Notes

The metrics could also be retrieved by the `stats/summary` endpoint.

### How to test
Run the unit test with `ddev test kubelet` and manually test as follows.

The volume metrics are not exposed by the `cadvisors/metrics` endpoint on kubelet when using containerd (See [upstream](https://github.com/google/cadvisor/issues/2785#issuecomment-1205538108)).
It is possible to retrieve them with docker as container runtime and kubernetes > v1.25.
You can test with the following procedure: 
- Start a minikube cluster with a docker runtime with : `minikube start -p="$CLUSTER_NAME" --container-runtime=docker`
- Build the image by writing the following Dockerfile in `DataDog/integrations-core`:
```dockerfile
FROM datadog/agent:latest
COPY kubelet/datadog_checks/kubelet/kubelet.py    /opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/kubelet/kubelet.py
```
and the command : `minikube -p="$CLUSTER_NAME" image build -t <image>:<tag> .`
- Deploy the agent and cluster agent with helm. I used:
```yaml
agents:
  image:
    repository: docker.io/datadog/agent
    tag: 7.44.0-rc.5 
```
- Deploy the following `manifest.yaml`:
```yaml
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: my-local-pv
spec:
  capacity:
    storage: 3Gi
  accessModes:
  - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: persistent-vol
  local:
    path: /data
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/hostname
          operator: In
          values:
          - ali-cluster
---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: persistent-vol
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim-demo-data
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: persistent-vol
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: my-ephemeral-pv
spec:
  capacity:
    storage: 3Gi
  accessModes:
  - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: ephemeral-sc
  local:
    path: /data
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/hostname
          operator: In
          values:
          - ali-cluster
---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: ephemeral-sc
provisioner: kubernetes.io/no-provisioner
volumeBindingMode: WaitForFirstConsumer

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: claim-demo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: claim-demo
  template:
    metadata:
      labels:
        app: claim-demo
        app.kubernetes.io/name: claim-demo
    spec:
      containers:
        - name: ubuntu
          image: ubuntu
          command: [ "sleep", "infinity" ]
          volumeMounts:
            - mountPath: /data
              name: data
          resources:
            requests:
              cpu: 50m
              memory: 100Mi
      terminationGracePeriodSeconds: 5
      volumes:
        - name: data
          persistentVolumeClaim:
            claimName: claim-demo-data
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ephemeral-demo
  labels:
    app: ephemeral-demo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: ephemeral-demo
  template:
    metadata:
      labels:
        app: ephemeral-demo
        app.kubernetes.io/name: ephemeral-demo
    spec:
      containers:
        - name: ubuntu
          image: ubuntu
          command: [ "sleep", "infinity" ]
          volumeMounts:
            - mountPath: "/scratch"
              name: scratch-volume
          resources:
            requests:
              cpu: 50m
              memory: 100Mi
      volumes:
        - name: scratch-volume
          ephemeral:
            volumeClaimTemplate:
              spec:
                accessModes: [ "ReadWriteOnce" ]
                storageClassName: ephemeral-sc
                resources:
                  requests:
                    storage: 1Gi
      terminationGracePeriodSeconds: 5
```
- Run `kubectl exec <agent pod name> -- agent check kubelet -r --table`

Before, tags are missing: 
![Capture d’écran 2023-03-21 à 15 23 17](https://user-images.githubusercontent.com/125997632/226643731-3bc7540b-29d6-4863-930f-1448199d2315.png)

After, tags are present: 
![Capture d’écran 2023-03-21 à 15 48 56](https://user-images.githubusercontent.com/125997632/226643922-14843cf7-fcf6-465d-a00f-65505f086dcf.png)


### Review checklist (to be filled by reviewers)

- [ ] If the PR doesn't need to be tested during QA, please add a `qa/skip-qa` label.
